### PR TITLE
Replace all instances of Guava cache with Caffeine

### PR DESCRIPTION
**Pull request by cshannon**
[Caffeine](https://github.com/ben-manes/caffeine) is the replacement for the Guava cache api and has several improvements including efficiency with its admission [policy](https://github.com/ben-manes/caffeine/wiki/Efficiency) and [performance](https://github.com/ben-manes/caffeine/wiki/Benchmarks). Currently the Accumulo code base has a mixture where some places use Guava and some places use Caffeine so it is not consistent. Switching to Caffeine is nearly a drop in replacement and this commit replaces all instances of Guava with Caffeine for consistency and also adds a checkstyle rule to prevent using Guava in the future.

From the Guava [docs](https://guava.dev/releases/31.1-jre/api/docs/com/google/common/cache/CacheBuilder.html) regarding Caffeine: 

> **Prefer [Caffeine](https://github.com/ben-manes/caffeine/wiki) over Guava's caching API**
**The successor to Guava's caching API is [Caffeine](https://github.com/ben-manes/caffeine/wiki). Its API is designed to make it a nearly drop-in replacement** -- though it requires Java 8 APIs, is not available for Android or GWT/j2cl, and may have [different (usually better) behavior](https://github.com/ben-manes/caffeine/wiki/Guava) when multiple threads attempt concurrent mutations. Its equivalent to CacheBuilder is its [Caffeine](https://www.javadoc.io/doc/com.github.ben-manes.caffeine/caffeine/latest/com.github.benmanes.caffeine/com/github/benmanes/caffeine/cache/Caffeine.html) class. **Caffeine offers better performance, more features (including asynchronous loading), and fewer [bugs](https://github.com/google/guava/issues?q=is%3Aopen+is%3Aissue+label%3Apackage%3Dcache+label%3Atype%3Ddefect).**

### CODE REVIEW

**Review by dlmarion — APPROVED** [review]
LGTM

**Comment by dlmarion** [comment]
Kicked off full IT build

**Comment by dlmarion** [comment]
Full IT build passed successfully

File: core/src/main/java/org/apache/accumulo/core/classloader/URLContextClassLoaderFactory.java [hunk 1]
```diff
@@ -63,19 +62,15 @@ public ClassLoader getClassLoader(String context) {
...
-    }
+    return classloaders.get(context, k -> {
+      LOG.debug("Creating URLClassLoader for context, uris: {}", context);
+      return new URLClassLoader(Arrays.stream(context.split(",")).map(url -> {
+        try {
+          return new URL(url);
+        } catch (MalformedURLException e) {
+          throw new RuntimeException(e);
+        }
+      }).collect(Collectors.toList()).toArray(new URL[] {}), ClassLoader.getSystemClassLoader());
```
Comment (DomGarguilo) [inline core/src/main/java/org/apache/accumulo/core/classloader/URLContextClassLoaderFactory.java:73]:
```suggestion
      }).toArray(URL[]::new), ClassLoader.getSystemClassLoader());
```
Outside of the scope and not changed in this PR but noticed that the collect step can be skipped here.

Comment (cshannon) [reply core/src/main/java/org/apache/accumulo/core/classloader/URLContextClassLoaderFactory.java:73]: Good catch, I will go ahead and fix that.

File: core/src/main/java/org/apache/accumulo/core/file/blockfile/impl/CachableBlockFile.java [hunk 2]
```diff
@@ -170,8 +169,14 @@ public static class Reader implements Closeable {
...
-        return fileLenCache.get(cacheId, lengthSupplier::get);
-      } catch (ExecutionException e) {
+        return fileLenCache.get(cacheId, k -> {
+          try {
+            return lengthSupplier.get();
+          } catch (IOException e) {
+            throw new UncheckedIOException(e);
+          }
+        });
+      } catch (Exception e) {
```
Comment (ctubbsii) [inline core/src/main/java/org/apache/accumulo/core/file/blockfile/impl/CachableBlockFile.java:179]: @cshannon Can this catch block be narrowed to just the UncheckedIOException that you're throwing internally, or in the worst case, catching all RTEs? I don't see what checked exceptions this needs to catch.

File: pom.xml [hunk 3]
```diff
@@ -1129,6 +1129,10 @@
                   <property name="format" value="&amp;quot; [+] &amp;quot;" />
                   <property name="message" value="Unnecessary concatenation of string literals" />
                 </module>
+                <module name="RegexpSinglelineJava">
+                  <property name="format" value="com.google.common.cache[.]" />
```
Comment (ctubbsii) [inline pom.xml:1133]: The pattern should be: `"com[.]google[.]common[.]cache[.]"

File: server/manager/src/main/java/org/apache/accumulo/manager/recovery/RecoveryManager.java [hunk 4]
```diff
@@ -144,8 +144,14 @@ private void initiateSort(String sortId, String source, final String destination
...
-      return existenceCache.get(path, () -> manager.getVolumeManager().exists(path));
-    } catch (ExecutionException e) {
+      return existenceCache.get(path, k -> {
+        try {
+          return manager.getVolumeManager().exists(path);
+        } catch (IOException e) {
+          throw new UncheckedIOException(e);
+        }
+      });
+    } catch (Exception e) {
```
Comment (ctubbsii) [inline server/manager/src/main/java/org/apache/accumulo/manager/recovery/RecoveryManager.java:154]: Again, a very broad catch.

**Review by ctubbsii — COMMENTED** [review]
After merge review for follow-up changes.

File: core/src/main/java/org/apache/accumulo/core/file/blockfile/impl/CachableBlockFile.java (hunk 2 above)
Comment (cshannon) [reply core/src/main/java/org/apache/accumulo/core/file/blockfile/impl/CachableBlockFile.java:179]: I'll take a look, I can create a follow up PR with requested changes.

File: pom.xml (hunk 3 above)
Comment (cshannon) [reply pom.xml:1133]: Good catch

**Pull request merged**